File: DFS/pathSum.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


# DFS+递归+递减
class Solution:
    def pathSum(self, root: TreeNode, sum: int):
        res, path = [], []

        def Recur(root: TreeNode, target: int):
            if not root:
                return
            path.append(root.val)
            target -= root.val
            if target == 0 and not root.left and not root.right:
                res.append(list(path))
            Recur(root.left, target)
            Recur(root.right, target)
            path.pop()

        Recur(root, sum)
        return res

File: DFS/pathSum.cpp
#include <stdc++.h>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// 路径不一定非得从二叉树的根节点或叶节点开始或结束，但是其方向必须向下(只能从父节点指向子节点方向)。
class Solution {
public:
    int ans = 0;
    int* spaces;
    int pathSum(TreeNode* root, int sum) {
        int dep = Depth(root);
        spaces = new int[dep];
        DFS(root, sum, 0);
    }

    void DFS(TreeNode* root, int tar, int index){
        if (root == nullptr)
            return;
        spaces[index] = root->val;
        int tmp = 0;
        for (int i = index; i > -1; --i){
            tmp += spaces[index];
            if (tmp == tar)
                ans++;
        }
        DFS(root->left, tar, index + 1);
        DFS(root->right, tar ,index + 1);
    }

    int Depth(TreeNode* root){
        if (root == nullptr)
            return 0;
        return max(Depth(root->left), Depth(root->right)) + 1;
    }
};